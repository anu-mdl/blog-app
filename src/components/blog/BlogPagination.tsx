'use client';

// TODO: fix pagination scroll bug

import { Button } from '@/components/ui/button';
import { ChevronLeft, ChevronRight } from 'lucide-react';

interface BlogPaginationProps {
  currentPage: number;
  totalPages: number;
  onPageChange: (page: number) => void;
}

export function BlogPagination({
  currentPage,
  totalPages,
  onPageChange
}: BlogPaginationProps) {
  const pages = Array.from({ length: totalPages }, (_, i) => i + 1);

  const getVisiblePages = () => {
    if (totalPages <= 5) return pages;

    if (currentPage <= 3) {
      return [...pages.slice(0, 5), 'ellipsis', totalPages];
    }

    if (currentPage >= totalPages - 2) {
      return [1, 'ellipsis', ...pages.slice(totalPages - 5)];
    }

    return [
      1,
      'ellipsis',
      currentPage - 1,
      currentPage,
      currentPage + 1,
      'ellipsis',
      totalPages
    ];
  };

  const visiblePages = getVisiblePages();

  return (
    <div className="flex items-center justify-center gap-1 mt-8">
      <Button
        variant="outline"
        size="icon"
        onClick={() => onPageChange(currentPage - 1)}
        disabled={currentPage === 1}
      >
        <ChevronLeft className="h-4 w-4" />
        <span className="sr-only">Previous page</span>
      </Button>

      {visiblePages.map((page, index) => {
        if (page === 'ellipsis') {
          return (
            <span
              key={`ellipsis-${index}`}
              className="px-3 py-2 text-muted-foreground"
            >
              ...
            </span>
          );
        }

        return (
          <Button
            key={`page-${page}`}
            variant={currentPage === page ? 'default' : 'outline'}
            size="icon"
            onClick={() => onPageChange(Number(page))}
          >
            {page}
          </Button>
        );
      })}

      <Button
        variant="outline"
        size="icon"
        onClick={() => onPageChange(currentPage + 1)}
        disabled={currentPage === totalPages}
      >
        <ChevronRight className="h-4 w-4" />
        <span className="sr-only">Next page</span>
      </Button>
    </div>
  );
}
