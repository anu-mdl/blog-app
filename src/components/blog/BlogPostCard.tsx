import { PostsRecordExtended } from '@/api/extended_types';
import { Avatar, AvatarFallback, AvatarImage } from '@/components/ui/avatar';
import { Badge } from '@/components/ui/badge';
import {
  Card,
  CardContent,
  CardFooter,
  CardHeader
} from '@/components/ui/card';
import { pb } from '@/lib/pb';
import { formatDate } from '@/lib/utils';
import { MessageSquare } from 'lucide-react';
import Image from 'next/image';
import Link from 'next/link';

export function BlogPostCard({
  post,
  featured = false
}: {
  post: PostsRecordExtended;
  featured?: boolean;
}) {
  const commentCount = 5;
  if (featured) {
    return (
      <Card className="overflow-hidden border-0 shadow-none md:grid md:grid-cols-2 md:gap-4 md:border md:shadow-sm">
        <div className="relative aspect-video overflow-hidden rounded-t-lg md:aspect-auto md:rounded-lg">
          <Image
            src={post.image ? pb.files.getURL(post, post.image) : '/image.svg'}
            width={1200}
            height={600}
            alt="blog post"
            className="h-full w-full object-cover transition-transform duration-300 hover:scale-105"
            priority
          />
        </div>
        <div className="flex flex-col justify-center p-6">
          <CardHeader className="p-0">
            <div className="flex items-center gap-2 mb-3">
              <Badge variant="secondary" className="px-2 py-0 text-xs">
                {post.category}
              </Badge>
              <span className="text-sm text-muted-foreground">
                {formatDate(post.created || '')}
              </span>
            </div>
            <Link href={`/blog/${post.id}`} className="group">
              <h2 className="text-2xl font-bold tracking-tight transition-colors group-hover:text-primary md:text-3xl">
                {post.title}
              </h2>
            </Link>
          </CardHeader>
          <CardContent className="p-0 pt-4">
            <p className="text-muted-foreground line-clamp-3">{post.excerpt}</p>
          </CardContent>
          <CardFooter className="p-0 pt-6">
            <div className="flex items-center justify-between w-full">
              <div className="flex items-center gap-2">
                <Avatar className="h-8 w-8">
                  <AvatarImage
                    src={
                      post.expand?.author?.avatar
                        ? pb.files.getUrl(
                            post.expand.author,
                            post.expand.author.avatar
                          )
                        : '/image.svg'
                    }
                    alt={post.expand?.author?.username}
                  />
                  <AvatarFallback>
                    {post.expand?.author?.username?.charAt(0) || '?'}
                  </AvatarFallback>
                </Avatar>
                <span className="text-sm font-medium">
                  {post.expand?.author?.username}
                </span>
              </div>
              <div className="flex items-center gap-4 text-sm text-muted-foreground">
                <div className="flex items-center gap-1">
                  <MessageSquare className="h-4 w-4" />
                  <span>{commentCount}</span>
                </div>
              </div>
            </div>
          </CardFooter>
        </div>
      </Card>
    );
  }

  return (
    <Card className="overflow-hidden h-full transition-all hover:border-primary/50">
      <div className="relative aspect-video overflow-hidden rounded-t-lg">
        <Image
          src={post.image ? pb.files.getURL(post, post.image) : '/image.svg'}
          alt="blog post"
          width={354}
          height={199}
          className="h-full w-full object-cover transition-transform duration-300 hover:scale-105"
        />
        <div className="absolute top-2 left-2">
          <Badge variant="secondary" className="px-2 py-0 text-xs">
            {post.category}
          </Badge>
        </div>
      </div>
      <CardHeader className="p-4 pb-0">
        <div className="text-sm text-muted-foreground mb-2">
          {formatDate(post.created || '')}
        </div>
        <Link href={`/blog/${post.id}`} className="group">
          <h3 className="font-bold tracking-tight transition-colors group-hover:text-primary line-clamp-2">
            {post.title}
          </h3>
        </Link>
      </CardHeader>
      <CardContent className="p-4 pt-2">
        <p className="text-muted-foreground text-sm line-clamp-2">
          {post.excerpt}
        </p>
      </CardContent>
      <CardFooter className="p-4 pt-0 mt-auto">
        <div className="flex items-center justify-between w-full">
          <div className="flex items-center gap-2">
            <Avatar className="h-6 w-6">
              <AvatarImage
                src={
                  post.expand?.author?.avatar
                    ? pb.files.getUrl(
                        post.expand.author,
                        post.expand.author.avatar
                      )
                    : '/image.svg'
                }
                alt={post.expand?.author?.username}
              />
              <AvatarFallback>
                {post.expand?.author?.username?.charAt(0) || '?'}
              </AvatarFallback>
            </Avatar>
            <span className="text-xs">{post.expand?.author?.username}</span>
          </div>
          <div className="flex items-center gap-3 text-xs text-muted-foreground">
            <div className="flex items-center gap-1">
              <MessageSquare className="h-3 w-3" />
              <span>{commentCount}</span>
            </div>
          </div>
        </div>
      </CardFooter>
    </Card>
  );
}
