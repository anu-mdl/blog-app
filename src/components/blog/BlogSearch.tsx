'use client';

import type React from 'react';

import { useState } from 'react';
import { Input } from '@/components/ui/input';
import { Button } from '@/components/ui/button';
import { Search } from 'lucide-react';

interface BlogSearchProps {
  onSearch: (query: string) => void;
}

export function BlogSearch({ onSearch }: BlogSearchProps) {
  const [query, setQuery] = useState('');

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();
    onSearch(query);
  };

  return (
    <form onSubmit={handleSubmit} className="relative">
      <Input
        type="search"
        placeholder="Search articles..."
        className="pr-10"
        value={query}
        onChange={e => setQuery(e.target.value)}
      />
      <Button
        type="submit"
        size="icon"
        variant="ghost"
        className="absolute right-0 top-0 h-full px-3 text-muted-foreground hover:text-foreground"
      >
        <Search className="h-4 w-4" />
        <span className="sr-only">Search</span>
      </Button>
    </form>
  );
}
