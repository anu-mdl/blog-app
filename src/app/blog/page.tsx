'use client';

import { Input } from '@/components/ui/input';

import { useState } from 'react';
import { BlogPostCard } from '@/components/blog/BlogPostCard';
import { BlogSearch } from '@/components/blog/BlogSearch';
import { BlogPagination } from '@/components/blog/BlogPagination';
import { Badge } from '@/components/ui/badge';
import { Button } from '@/components/ui/button';
import { Tabs, TabsContent, TabsList, TabsTrigger } from '@/components/ui/tabs';
import { cn } from '@/lib/utils';
import { pb } from '@/lib/pb';
import { useQuery } from '@tanstack/react-query';
import { PostsRecordExtended } from '@/api/extended_types';
import { useRouter, useSearchParams } from 'next/navigation';

const categories = [
  'All',
  'Technology',
  'React',
  'Next.js',
  'CSS',
  'Accessibility',
  'Security'
];

export default function BlogPage() {
  const [currentPage, setCurrentPage] = useState(1);
  const [activeCategory, setActiveCategory] = useState('All');
  // const [searchQuery, setSearchQuery] = useState('');
  const searchParams = useSearchParams();
  const searchQuery = searchParams.get('search') || '';
  const router = useRouter();

  const { data, isLoading, error } = useQuery({
    queryKey: ['blog-posts'],
    queryFn: async () => {
      const response = await pb
        .collection('posts')
        .getFullList<PostsRecordExtended>({
          sort: 'created',
          expand: 'author'
        });
      return response;
    }
  });

  if (isLoading) return <div>Loading...</div>;
  if (error) return <div>Error loading data</div>;
  if (!data || data.length === 0) return <div>No data available</div>;

  const postsPerPage = 6;

  const filteredPosts = data?.filter(post => {
    const matchesCategory =
      activeCategory === 'All' || post.category === activeCategory;
    const matchesSearch =
      post.title?.toLowerCase().includes(searchQuery.toLowerCase()) ||
      post.excerpt?.toLowerCase().includes(searchQuery.toLowerCase()) ||
      post.tags?.toLowerCase().includes(searchQuery.toLowerCase());
    return matchesCategory && matchesSearch;
  });

  const handleClearSearch = () => {
    const params = new URLSearchParams(searchParams.toString());
    params.delete('search');
    router.push(`?${params.toString()}`);
  };

  const indexOfLastPost = currentPage * postsPerPage;
  const indexOfFirstPost = indexOfLastPost - postsPerPage;
  const currentPosts = filteredPosts?.slice(indexOfFirstPost, indexOfLastPost);
  const totalPages = Math.ceil(filteredPosts.length / postsPerPage);

  const handlePageChange = (pageNumber: number) => {
    setCurrentPage(pageNumber);
    window.scrollTo({ top: 0, behavior: 'smooth' });
  };

  const handleSearch = (query: string) => {
    // setSearchQuery(query);
    const params = new URLSearchParams(searchParams.toString());
    params.set('search', query);
    router.push(`?${params.toString()}`);
    setCurrentPage(1);
  };

  const handleCategoryChange = (category: string) => {
    setActiveCategory(category);
    setCurrentPage(1);
  };

  const featuredPost = data[0];

  return (
    <div className="container mx-auto px-4 py-8 md:py-12">
      {/* HEADER */}
      <div className="mb-8 md:mb-12 text-center">
        <h1 className="text-3xl md:text-5xl font-bold tracking-tight mb-4">
          Some Blog
        </h1>
        <p className="text-muted-foreground text-lg max-w-2xl mx-auto">
          Discover the latest insights and news about web development, design,
          and technology.
        </p>
      </div>

      <div className="grid grid-cols-1 md:grid-cols-[1fr_300px] gap-8 mb-8">
        <div className="space-y-8">
          {/* SEARCH BAR */}
          <BlogSearch onSearch={handleSearch} />
          {searchQuery && (
            <div className="flex items-center gap-2">
              <span className="text-sm text-muted-foreground">
                Search results for:
              </span>
              <Badge variant="secondary" className="text-xs">
                {searchQuery}
                <Button
                  variant="ghost"
                  size="icon"
                  className="h-4 w-4 ml-1 p-0"
                  // onClick={() => setSearchQuery('')}
                  onClick={() => handleClearSearch()}
                >
                  <span className="sr-only">Clear search</span>Ã—
                </Button>
              </Badge>
            </div>
          )}
          {!searchQuery && <BlogPostCard post={featuredPost} featured={true} />}
        </div>

        {/* CATEGORIES FILTER*/}
        <div className="space-y-6">
          <div className="bg-muted/40 p-6 rounded-lg">
            <h2 className="font-semibold mb-4">Categories</h2>
            <div className="flex flex-wrap gap-2">
              {categories.map(category => (
                <Badge
                  key={category}
                  variant={activeCategory === category ? 'default' : 'outline'}
                  className={cn(
                    'cursor-pointer hover:bg-primary/20',
                    activeCategory === category && 'hover:bg-primary'
                  )}
                  onClick={() => handleCategoryChange(category)}
                >
                  {category}
                </Badge>
              ))}
            </div>
          </div>

          {/* TODO: add tags filter??? */}

          <div className="bg-muted/40 p-6 rounded-lg">
            <h2 className="font-semibold mb-4">Newsletter</h2>
            <p className="text-sm text-muted-foreground mb-4">
              Subscribe to our newsletter to get the latest updates directly in
              your inbox.
            </p>
            <form className="space-y-2">
              <Input type="email" placeholder="Your email address" />
              <Button className="w-full" disabled={true}>
                Subscribe
              </Button>
            </form>
          </div>
        </div>
      </div>

      {/* TABS */}
      <Tabs defaultValue="latest" className="mb-8">
        <div className="flex items-center justify-between border-b">
          <TabsList className="h-auto p-0">
            <TabsTrigger
              value="latest"
              className="data-[state=active]:border-b-2 data-[state=active]:border-primary rounded-none px-4 py-2 -mb-px data-[state=active]:text-foreground"
            >
              Latest
            </TabsTrigger>
            <TabsTrigger
              value="trending"
              className="data-[state=active]:border-b-2 data-[state=active]:border-primary rounded-none px-4 py-2 -mb-px data-[state=active]:text-foreground"
            >
              Trending
            </TabsTrigger>
            {/* <TabsTrigger
                value="popular"
                className="data-[state=active]:border-b-2 data-[state=active]:border-primary rounded-none px-4 py-2 -mb-px data-[state=active]:text-foreground"
              >
                Popular
              </TabsTrigger> */}
          </TabsList>

          <div className="text-sm text-muted-foreground">
            {filteredPosts.length}{' '}
            {filteredPosts.length === 1 ? 'article' : 'articles'}
          </div>
        </div>

        {/* POSTS TABS*/}
        <TabsContent value="latest" className="pt-6">
          {currentPosts.length > 0 ? (
            <div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-6">
              {currentPosts.map(post => (
                <BlogPostCard key={post.id} post={post} />
              ))}
            </div>
          ) : (
            <div className="text-center py-12">
              <h3 className="text-lg font-medium mb-2">No articles found</h3>
              <p className="text-muted-foreground">
                Try adjusting your search or filter to find what you&apos;re
                looking for.
              </p>
            </div>
          )}
        </TabsContent>

        <TabsContent value="trending" className="pt-6">
          <div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-6">
            {[...currentPosts]
              .sort((a, b) => (b.likes || 0) - (a.likes || 0))
              .map(post => (
                <BlogPostCard key={post.id} post={post} />
              ))}
          </div>
        </TabsContent>

        {/* TODO: make sort by comments count */}
        {/* <TabsContent value="popular" className="pt-6">
          <div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-6">

          </div>
        </TabsContent> */}
      </Tabs>

      {/* PAGINATION */}
      {totalPages > 1 && (
        <BlogPagination
          currentPage={currentPage}
          totalPages={totalPages}
          onPageChange={handlePageChange}
        />
      )}
    </div>
  );
}
