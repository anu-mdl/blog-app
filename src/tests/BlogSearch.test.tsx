import { BlogSearch } from '@/components/blog/BlogSearch';
import renderer from 'react-test-renderer';

jest.mock('@/components/ui/input', () => (props: any) => (
  <input {...props} data-testid="mock-input" />
));
jest.mock('@/components/ui/button', () => (props: any) => (
  <button {...props} data-testid="mock-button" />
));
jest.mock('lucide-react', () => ({
  Search: (props: any) => <svg {...props} data-testid="mock-search-icon" />
}));

describe('BlogSearch Component', () => {
  it('renders correctly and matches snapshot', () => {
    const mockOnSearch = jest.fn();
    const tree = renderer
      .create(<BlogSearch onSearch={mockOnSearch} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with an initial query (if you want to test different states)', () => {
    const mockOnSearch = jest.fn();
    const component = renderer.create(<BlogSearch onSearch={mockOnSearch} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
