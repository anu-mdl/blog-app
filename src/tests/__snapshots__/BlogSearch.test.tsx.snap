// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlogSearch Component renders correctly and matches snapshot 1`] = `null`;

exports[`BlogSearch Component renders correctly with an initial query (if you want to test different states) 1`] = `null`;
